File: covid19.r
library(rvest)
library(tidyverse)
library(reshape2)
library(ggplot2)
library(plotly)
library(viridis)
library(curl)
library(emayili)
library(orca)
library(telegram.bot)

g_username <- 'youremail@domain.com'
g_password <- 'password'
g_port <- 587
g_host <- 'smtp.gmail.com'
to_email <- 'toxxxxx@domain.com'
bot_token <- 'xxxxxxxxxxxxx:xxxxxxxxxxxx-xxxxxxxxxxxxxxxxxx'
bot_chat_id <- 'xxxxxxxxxxxxxx'


# GET DATA FROM WORDOMETER
getData <- function(){
  options(stringsAsFactors = FALSE)
  
  url_base <- "https://www.worldometers.info/coronavirus/"
  corona <- data.frame(
    "Country" = character(0),
    "Total_Cases" = character(0),
    "New_Cases" = character(0),
    "Total_Deaths" = integer(0),
    "New_Deaths" = character(0),
    "Total_Recovered" = character(0),
    "Active_Cases" = character(0),
    "Serious_Critical" = character(0),
    "Tot_Cases_1M_pop" = character(0))
  
  
  url <- paste0(url_base)
  pageTable <- url %>%
    read_html() %>%
    html_nodes(xpath='//*[@id="main_table_countries"]') %>%
    html_table()
  names(pageTable[[1]]) <- names(corona)
  corona <- rbind(corona, pageTable[[1]])
  
  # Remove comma characters in the columns
  corona$Total_Cases <- as.numeric(gsub(',','',corona$Total_Cases))
  corona$Total_Recovered <- as.numeric(gsub(',','',corona$Total_Recovered))
  corona$Total_Deaths <- as.numeric(gsub(',','',corona$Total_Deaths))
  corona$New_Cases <- as.numeric(gsub(',+','',corona$New_Cases))
  corona$Active_Cases <- as.numeric(gsub(',+','',corona$Active_Cases))
  corona$Serious_Critical <- as.numeric(gsub(',+','',corona$Serious_Critical))
  plotChart(corona)
  sendEmail(corona)
  sendTelegramBot(corona)
  delFiles()
  
  
}
  
# PLOT GRAPHS
plotChart <- function(corona){
  # plot the first 10 countries with the highest number of cases
  plot10 <- head(corona,10)
  plotMelt<-plot10 %>% select(Country,Total_Cases,Total_Recovered,Total_Deaths,Serious_Critical) %>% melt(id.vars="Country")
  plotNotMelt<-plot10 %>% select(Country,Total_Cases,Total_Recovered,Total_Deaths,Serious_Critical)
  
  
  p <- ggplot(plotMelt, aes(fill=variable, y=value, x=Country)) +
    geom_bar(position="dodge", stat="identity") +
    scale_fill_viridis(discrete = T) +
    ggtitle("CORONA VIRUS") +
    theme_minimal() +
    xlab("")
  p
  # fig <- plot_ly(plotNotMelt, x = ~Country, y = ~Total_Cases, type = 'bar', name = 'Total Cases')
  # fig <- fig %>% add_trace(y = ~Total_Recovered, name = 'Total Recovered')
  # fig <- fig %>% add_trace(y = ~Total_Deaths, name = 'Total Deaths')
  # fig <- fig %>% add_trace(y = ~Serious_Critical, name = 'Serious Critical')
  # fig <- fig %>% layout(yaxis = list(title = 'Number Of Cases'), barmode = 'group',title ="CORONA PANDEMIC")
  # fig
  
  write.csv(corona, 'coronaV.csv')
  ggsave(plot = p,"coronaV.png",scale = 1)
  
}
  
  
  # SENDING EMAIL FUNCTION
  sendEmail <- function(corona){
    
    msg <- paste0(
      "<h2>Hello Mary</h2> <br>",
      "<b style=color:red>Total Cases:&nbsp;",last(corona$Total_Cases),"</b> <br>",
      "<b style=color:red >Total Deaths:&nbsp;",last(corona$Total_Deaths),"</b> <br>",
      "<b style=color:purple>Recovered Cases:&nbsp;",last(corona$Total_Recovered),"</b> <br>",
      "<b style =color:red>Critical Serious:&nbsp;",last(corona$Serious_Critical),"</b> <br>"
    )
    
    email <- envelope() %>%
      from(g_username) %>%
      to(to_email) %>%
      subject("ℍ𝔼ℝ𝔼 𝕀𝕊 𝕋𝕆𝔻𝔸𝕐𝕊 ℂ𝕆ℝ𝕆ℕ𝔸 𝕍𝕀ℝ𝕌𝕊 𝕌ℙ𝔻𝔸𝕋𝔼 [ ℂ𝕆𝕍𝕀𝔻𝟙𝟡] 👹") %>% 
      html(msg) %>% 
      attachment("coronaV.png") %>% 
      attachment("coronaV.csv")
    
    smtp <- server(host = g_host,
                   port = g_port,
                   username = g_username,
                   password = g_password)
    
    smtp(email, verbose = TRUE)
    
  }
  
  
  sendTelegramBot <- function(corona){
    bot = Bot(token = bot_token)
    msg <- paste0("Hello John",
                  " ℍ𝔼ℝ𝔼 𝕀𝕊 𝕋𝕆𝔻𝔸𝕐𝕊 ℂ𝕆ℝ𝕆ℕ𝔸 𝕍𝕀ℝ𝕌𝕊 𝕌ℙ𝔻𝔸𝕋𝔼 [ ℂ𝕆𝕍𝕀𝔻𝟙𝟡] 👹 ",
                  "Total Cases:",last(corona$Total_Cases),

                  "Total Deaths:",last(corona$Total_Deaths),

                  "Recovered Cases:",last(corona$Total_Recovered),
                  "Critical Serious:",last(corona$Serious_Critical)
    )

    bot$sendMessage(chat_id = bot_chat_id, text = msg)
    bot$sendPhoto(chat_id = bot_chat_id, photo = 'coronaV.png')
    bot$sendDocument(chat_id = bot_chat_id, document = 'coronaV.csv')


  }
  
  
  # DELETE FILES GENERATED FOR NEXT NOTIFICATION
  delFiles <- function(){
    fn <- c('coronaV.png','coronaV.csv')
    for (n in fn){
      if (file.exists(n)) 
      file.remove(n)
    }
    print("FILES DELETED !!!")
  }
  
  
  
  getData()
  
  

  #   # usr/bin/Rscript --vanilla coronavirus.R                                     
  
  
 

File: covid19.py
import requests
import os
import pandas as pd
import smtplib
from lxml import html
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from email.mime.base import MIMEBase
from email import encoders
import seaborn as sns
import matplotlib.pyplot as plt
from sendgrid import SendGridAPIClient
from sendgrid.helpers.mail import (Mail, Attachment, FileContent, FileName, FileType, Disposition,To)
import base64
import mimetypes
from twilio.rest import Client
import telebot
from matplotlib.backends.backend_pdf import PdfPages
import time
import slack







# you can set cron job for linux like this
# * * * * * /usr/bin/python3  /to/path coronavirus.py >/dev/null 2>&1


class Coronavirus():
    def __init__(self):
       self.sender    = "sender@email.com"

       self.receiver  = [('receiver@email.com','John doe'),('another_receiver@email.com','Mary jane')]
       self.g_receiver = 'greceiver@email.com' #if using google

       self.to_emails = [ To( email='John@John.com' ,
        name='John Doe',
        substitutions={
        '-name-': 'John',
        },
        subject='Hi -name-, Here is today"s Corona Virus Update [ Covid19]' 

        ),

        To(email='Mary.b.Mary@Mary.com',
        name='Mary',
        substitutions = {
        '-name-' : 'Mary'
        }
        )
        ]

       self.password  = "password" # password if using  google's gmail
       self.attachments = ['coronaV.png', 'coronaV.csv','coronaV.pdf'] # attachments
       self.api = 'SG.xxxxx.xxxxxxx-xxxxx' #sendgrid api
       self.account_sid = "xxxxxxxxxxxxxxxxxxxxxxxxxx"
       self.auth_token = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"

    #    twilio whats app
       self.to_whatsapp = [('+1xxxxxxxxxxxxxxxx','Mary'),('+xxxxxxxxxxxxxxx','John')]
       self.from_whatsapp = '+xxxxxxxxxxxxxxxxxxxx'

    #    telegram api token and chat id
       self.bot_token = 'xxxxxxxxxxxxxxxx:xxxxxxxxxxxxxxx-xxxxxxxxxx'
       self.bot_chatID = '654564773'
    
     #    slack token
       self.SLACK_API_TOKEN ='xoxb-xxxxxxxxxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxxxx-xxxxx'


    # scrape worldometers
    def getCoronaData(self):
        url    = 'https://www.worldometers.info/coronavirus/'
        htmly  = requests.get(url).content
        htmlyr = html.fromstring(htmly)

        df_list = pd.read_html(htmly)
        Tcases = htmlyr.xpath('//*[@id="maincounter-wrap"]/div/span/text()')[0][:]
        TDeaths = htmlyr.xpath('//*[@id="maincounter-wrap"]/div/span/text()')[1][:]
        Recovered = htmlyr.xpath('//*[@id="maincounter-wrap"]/div/span/text()')[2][:]

        df = df_list[0]
        print(df)
        df.to_csv('coronaV.csv')
        pp = PdfPages("coronaV.pdf")
        pp.close()

        self.Plots(df)
        self.SendGrid(Tcases,TDeaths,Recovered)
        # self.sendgrid()
        self.notifyWhatsapp(Tcases,TDeaths,Recovered)
        self.notifyTelegram(Tcases,TDeaths,Recovered)
        self.notifySlack(Tcases,TDeaths,Recovered)
        self.DelFiles()


    # Send Notification emails with sendgrid
    def SendGrid(self,Tcases,TDeaths,Recovered):
         # # Add body to email
        body = """ <html>
                   <body>
                        <u><strong>COVID19<strong></u><br><br>
                            <b>Total Cases: {Tcases}  </b><br>
                            <b style="color:red">Total Deaths : {TDeaths}  </b><br>
                            <b>Recovered : {Recovered}  </b><br>
                            <a href="https://www.worldometers.info/coronavirus/" >Check the link: click here </a>
                    </body>
                    </html>
                """.format(Tcases=Tcases,TDeaths=TDeaths,Recovered=Recovered)


        message = Mail(
        from_email= (self.sender,'John Doe'),
        to_emails= self.receiver,
        subject='Corona Virus Update [ Covid19] [Local]',
        html_content=body,
        is_multiple= True)

        for filename in self.attachments:
            with open(filename, 'rb') as f:
                data = f.read()
                f.close()
                encoded_file = base64.b64encode(data).decode()

                attachedFile = Attachment(
                    FileContent(encoded_file),
                    FileName(filename),
                    FileType(mimetypes.guess_type(filename)[0]),
                    Disposition('attachment')
                )
                print("filename =%s " % filename)
                print(mimetypes.guess_type(filename))
                message.attachment = attachedFile

        try:
            sg = SendGridAPIClient(self.api)
            response = sg.send(message)
            print(response.status_code)
            print(response.body)
            print(response.headers)
        except Exception as e:
            print(e)

        
    # Send emails using googles gmail
    def SendEmail(self,Tcases,TDeaths,Recovered):
        msg = MIMEMultipart()

        msg['Subject'] = 'Corona Virus Update [ Covid19]'
        msg['From'] = self.sender
        msg['To'] =   self.g_receiver
        
        # # Add body to email
        body = 'Total Cases: ' + Tcases + '\
        \nTotal Deaths : ' + TDeaths + '\
        \nRecovered : ' + Recovered + '\
        \nCheck the link: https://www.worldometers.info/coronavirus/'
        msg.attach(MIMEText(body,"plain"))
            
        
        if  len(self.attachments) > 0: # are there attachments?
            
            for filename in self.attachments:
                f = filename
                part = MIMEBase('application', "octet-stream")
                part.set_payload( open(f,"rb").read() )
                encoders.encode_base64(part)
                part.add_header('Content-Disposition', 'attachment; filename="%s"' % os.path.basename(f))
                msg.attach(part)


        
        user = self.sender
        password = self.password
        
        server = smtplib.SMTP('smtp.gmail.com', 587)
        server.connect("smtp.gmail.com",587)
        server.ehlo()
        server.starttls()
        server.ehlo()
        server.login(user, password)
        server.sendmail(self.sender, self.receiver, msg.as_string())
        print("Successfully sent email")
        server.quit()

       
        
    def Plots(self,df):
        df2 = df[['Country,Other','TotalCases','TotalDeaths','TotalRecovered','Serious,Critical']]
        df2 = df2.head(10)
        
        df2= pd.melt(df2, id_vars=['Country,Other']).sort_values(['variable','value'])
        
        g = sns.catplot(x="Country,Other", y="value", hue="variable", data=df2,height=6, kind="bar", palette = 'twilight',aspect=1.5)
        g.despine(left=True,bottom=True)
        g.set_xlabels("Countries")
        g.set_ylabels("Cases in thousands")
        plt.xticks(rotation=90)
        plt.title('Ten(10) With Most covid19 cases')
        g.savefig("coronaV.png")
        
    
    def DelFiles(self):
        files = ['coronaV.png', 'coronaV.csv','coronaV.pdf']
        for f in files:
            os.remove(f)


    
    
    def notifyTelegram(self,Tcases,TDeaths,Recovered):


        message ='ℍ𝔼ℝ𝔼 𝕀𝕊 𝕋𝕆𝔻𝔸𝕐𝕊 ℂ𝕆ℝ𝕆ℕ𝔸 𝕍𝕀ℝ𝕌𝕊 𝕌ℙ𝔻𝔸𝕋𝔼 [ ℂ𝕆𝕍𝕀𝔻𝟙𝟡] 👹 \
        \nTotal Cases: ' + Tcases + '\
        \nTotal Deaths : ' + TDeaths + '\
        \nRecovered : ' + Recovered + '\
        \nCheck the link: https://www.worldometers.info/coronavirus/'


        send_text = 'https://api.telegram.org/bot' + self.bot_token + '/sendMessage?chat_id=' + self.bot_chatID + '&parse_mode=Markdown&text=' + message

        requests.get(send_text)

        # url = f"https://api.telegram.org/bot{self.bot_token}/sendPhoto"
        # files = {'photo': open('coronaV.png', 'rb')}
        # data = {'chat_id' : self.bot_chatID}
        # r= requests.post(url, files=files, data=data)
        # print(r.status_code, r.reason, r.content)

        bot = telebot.TeleBot(self.bot_token)
        # doc = open('book.csv', 'rb')
        # bot.send_document(self.bot_chatID, doc)
        # bot.send_document(self.bot_chatID, "FILEID")




        for media in self.attachments:
            ext = os.path.splitext(media)[-1].lower()
            print(ext)

            if ext == ".pdf":
                doc = open(media, 'rb')
                bot.send_document(self.bot_chatID, doc)


            if ext == ".png": 
                doc = open(media, 'rb')
                bot.send_photo(self.bot_chatID, doc)
        print("Telegram Notification Sent!")


            
        

    
    def notifyWhatsapp(self,Tcases,TDeaths,Recovered):

        client = Client(self.account_sid,self.auth_token)
        from_whatsapp_number = 'whatsapp:'+self.from_whatsapp
        to_whatsapp_number = self.to_whatsapp

        if len(to_whatsapp_number) >0:
            for num in to_whatsapp_number:
                client.messages.create(
                            body=""" Hi {jina},  ℍ𝔼ℝ𝔼 𝕀𝕊 𝕋𝕆𝔻𝔸𝕐𝕊 ℂ𝕆ℝ𝕆ℕ𝔸 𝕍𝕀ℝ𝕌𝕊 𝕌ℙ𝔻𝔸𝕋𝔼 [ ℂ𝕆𝕍𝕀𝔻𝟙𝟡] 👹 

                𝕿𝖔𝖙𝖆𝖑 𝕮𝖆𝖘𝖊𝖘 : {Tcases} 
                𝕿𝖔𝖙𝖆𝖑 𝕯𝖊𝖆𝖙𝖍𝖘 : {TDeaths}
                𝕽𝖊𝖈𝖔𝖛𝖊𝖗𝖊𝖉  : {Recovered} 
                              """.format(jina = num[1] ,Tcases=Tcases,TDeaths=TDeaths,Recovered=Recovered) ,
                            from_=from_whatsapp_number,
                            to='whatsapp:'+num[0])
        print("Whatsapp Notification Sent!")
        
        
      
    
     def notifySlack(self,Tcases,TDeaths,Recovered):
        client = slack.WebClient(token=self.SLACK_API_TOKEN)
        body=""" Hi,  ℍ𝔼ℝ𝔼 𝕀𝕊 𝕋𝕆𝔻𝔸𝕐𝕊 ℂ𝕆ℝ𝕆ℕ𝔸 𝕍𝕀ℝ𝕌𝕊 𝕌ℙ𝔻𝔸𝕋𝔼 [ ℂ𝕆𝕍𝕀𝔻𝟙𝟡] 👹 
        𝕿𝖔𝖙𝖆𝖑 𝕮𝖆𝖘𝖊𝖘 : {Tcases} 
        𝕿𝖔𝖙𝖆𝖑 𝕯𝖊𝖆𝖙𝖍𝖘 : {TDeaths}
        𝕽𝖊𝖈𝖔𝖛𝖊𝖗𝖊𝖉    : {Recovered}""".format(Tcases=Tcases,TDeaths=TDeaths,Recovered=Recovered)

        response = client.chat_postMessage(
            channel='#diy-projects',
            text=body)
        assert response["ok"]

        for media in self.attachments:
            resUpload = client.files_upload(
            channels='#diy-projects',
            file=media)
            assert resUpload["ok"]
        print("Slack Notification Sent!")


            
        


Coronavirus().getCoronaData()
        
